Port radxa cm5 io from linux-6.1-stan-rkr4.1

nascs opened the pull request.

nascs commented:
1. SD Card/eMMC/USB 可以启动， 其中 3 个 USB 口都能启动
2. 从 label 名为 config 或者 CONFIG 的分区加载 bmp

nascs commented:
已知 bug: 

Bus usb@fcd00000: Failed to get USB PHY0 for usb@fcd00000
Port not available.

这个是由于 usb@fcd00000 的 phy 驱动和另外一个 phy 驱动某些函数有冲突， 但是不影响 usb 启动

RadxaYuntian commented:
> 这个是由于 usb@fcd00000 的 phy 驱动和另外一个 phy 驱动某些函数有冲突， 但是不影响 usb 启动

这是OTG口吧？

RadxaYuntian commented on configs/radxa-cm5-io-rk3588s_defconfig:
先用这个defconfig生成`.config`后用`make savedefconfig`重新生成一份

RadxaYuntian commented on arch/arm/mach-rockchip/radxa_img.c:
```diff
@@ -1,55 +1,145 @@
+		return -EINVAL;
+	}
+
+	if (!memcmp(buffer + DOS_FS_TYPE_OFFSET, "FAT", 3)) {
```
https://blog.csdn.net/edsam49/article/details/5893068
> uboot下支持fat12、16、32。在 fat_register_device这个接口中有一个bug，用mkdosfs格式化的SD卡，其生成的MBR信息跟在PC上不一样。fat.c中的代码片段&buffer[DOS_FS_TYPE_OFFSET]起始的3个字节数据是否为“FAT”来判断其文件系统是不是FAT格式的，DOS_FS_TYPE_OFFSET为0x36，但是用mkdosfs格式化的SD卡，其生成的MBR信息，其"FAT"描述在偏移0x52处，所以在fat_register_device（）接口中需要做兼容性调整。这个小bug可以从static int read_bootsectandvi(boot_sector *bs, volume_info *volinfo, int *fatsize)这个接口的实现里看出。在这个 read_bootsectandvi(boot_sector *bs, volume_info *volinfo, int *fatsize)接口中已经做了兼容，在boot sector中fat_length，如果为0的话，就做了增加偏移量，用mkdosfs格式化的SD卡，其生成的MBR信息中fat_length恰恰为0.

这个判断是你自己写的么？

nascs replied:
来自 https://github.com/radxa/u-boot/blob/b42d9e1bc49c318a02c81ba7c93b28319caa5ea0/fs/fat/fat.c#L86

RadxaYuntian commented on arch/arm/mach-rockchip/radxa_img.c:
```diff
@@ -1,55 +1,145 @@
-		free(buf);
-		return -ENOMEM;
-	}
+		if ((strncmp(volume_label, "config", 6) == 0) || (strncmp(volume_label, "CONFIG", 6) == 0)) {
```
```suggestion
		if (stricmp(volume_label, "config") == 0) {
```
你上面读取label的时候有末尾填0就应该可以不用指定字符串长度。这样可以保证比如`config1`不会匹配为成功

nascs replied:
u-boot 中没有 stricmp 的实现

RadxaYuntian commented on arch/arm/mach-rockchip/radxa_img.c:
```diff
@@ -1,55 +1,145 @@
 
-	if (fat_set_blk_dev(desc, &part) != 0) {
-		printf("Failed to set block device for FAT\n");
+	part_count = get_partition_count(desc);
```
这个函数就是跑part_get_info，后面也用了，感觉这边的调用就比较多余。不如在后面单独在失败的时候检查part_num是不是1，是1就return

RadxaYuntian commented on common/spl/spl_ab.c:
这是在干什么

RadxaYuntian reviewed (changes requested):
另外你用bsp的rk2410在你的分支上编译试试，看看现在支持的几个设备还能不能正常编译

nascs commented:
> > 这个是由于 usb@fcd00000 的 phy 驱动和另外一个 phy 驱动某些函数有冲突， 但是不影响 usb 启动
> 
> 这是OTG口吧？

usb3-host

nascs commented:
> 另外你用bsp的rk2410在你的分支上编译试试，看看现在支持的几个设备还能不能正常编译

会有一些问题的： 如：

arch/arm/dts/rk3588-rock-5-itx.dtb: ERROR (phandle_references): Reference to non-existent node or label "vdd_ddr_s0"

arch/arm/dts/rk3588-rock-5-itx.dtb: ERROR (phandle_references): Reference to non-existent node or label "vdd_log_s0"

nascs commented:
> > 另外你用bsp的rk2410在你的分支上编译试试，看看现在支持的几个设备还能不能正常编译
> 
> 会有一些问题的： 如：
> 
> arch/arm/dts/rk3588-rock-5-itx.dtb: ERROR (phandle_references): Reference to non-existent node or label "vdd_ddr_s0"
> 
> arch/arm/dts/rk3588-rock-5-itx.dtb: ERROR (phandle_references): Reference to non-existent node or label "vdd_log_s0"

昨天试了一下， 把 5a 的直接挪过去， 然后添加一个 rk3588s-rock-5a-u-boot.dtsi, 不会有设备树内容上的报错， 但是会有设备树大小的报错， 如：

Error at 'fdt_find_regions': FDT_ERR_NOSPACE

RadxaYuntian commented:
报错就不能把上下文也发出来么？是不是还有Failed to hash configuration的错误？

RadxaYuntian commented:
另外为什么5a有报错cm5就没有？

nascs commented:
> 报错就不能把上下文也发出来么？是不是还有Failed to hash configuration的错误？

  LD      lib/built-in.o
  CC      examples/standalone/hello_world.o
  CC      examples/standalone/stubs.o
  CC      examples/standalone/rkspi.o
  LD      examples/standalone/libstubs.o
  LD      examples/standalone/hello_world
  OBJCOPY examples/standalone/hello_world.srec
  OBJCOPY examples/standalone/hello_world.bin
  LD      examples/standalone/rkspi
aarch64-linux-gnu-ld.bfd: warning: cannot find entry symbol rkspi; defaulting to 000000000c100000
  OBJCOPY examples/standalone/rkspi.srec
  OBJCOPY examples/standalone/rkspi.bin
  LD      u-boot
  DTC     arch/arm/dts/rk3588s-rock-5a.dtb
  FDTGREP dts/dt.dtb
  FDTGREP dts/dt-spl.dtb
  FDTGREP dts/dt-tpl.dtb
Error at 'fdt_find_regions': FDT_ERR_NOSPACE
Error: FDT_ERR_BADMAGIC
make[2]: *** [dts/Makefile:63: dts/dt-tpl.dtb] Error 1
make[1]: *** [Makefile:892: dts/dt.dtb] Error 2
make: *** [Makefile:468: __build_one_by_one] Error 2
make: Leaving directory '/home/william/pro/bsp/.src/u-boot'
bsp

RadxaYuntian reviewed (commented):
`include/dt-bindings`里面的内容应该没有太大影响，其他设备树编译不过应该主要还是dtsi的改动比较大。要不把linux的相关dtsi以及你的rk3588-u-boot.dtsi单独放到dts/rockchip里面，然后板级设备树只改动一下include路径？

nascs commented:
> `include/dt-bindings`里面的内容应该没有太大影响，其他设备树编译不过应该主要还是dtsi的改动比较大。要不把linux的相关dtsi以及你的rk3588-u-boot.dtsi单独放到dts/rockchip里面，然后板级设备树只改动一下include路径？

rk3588-linux.dtsi 有些节点会和 rk3588-u-boot.dtsi 冲突

nascs commented:
> > 报错就不能把上下文也发出来么？是不是还有Failed to hash configuration的错误？
> 
> LD lib/built-in.o CC examples/standalone/hello_world.o CC examples/standalone/stubs.o CC examples/standalone/rkspi.o LD examples/standalone/libstubs.o LD examples/standalone/hello_world OBJCOPY examples/standalone/hello_world.srec OBJCOPY examples/standalone/hello_world.bin LD examples/standalone/rkspi aarch64-linux-gnu-ld.bfd: warning: cannot find entry symbol rkspi; defaulting to 000000000c100000 OBJCOPY examples/standalone/rkspi.srec OBJCOPY examples/standalone/rkspi.bin LD u-boot DTC arch/arm/dts/rk3588s-rock-5a.dtb FDTGREP dts/dt.dtb FDTGREP dts/dt-spl.dtb FDTGREP dts/dt-tpl.dtb Error at 'fdt_find_regions': FDT_ERR_NOSPACE Error: FDT_ERR_BADMAGIC make[2]: *** [dts/Makefile:63: dts/dt-tpl.dtb] Error 1 make[1]: *** [Makefile:892: dts/dt.dtb] Error 2 make: *** [Makefile:468: __build_one_by_one] Error 2 make: Leaving directory '/home/william/pro/bsp/.src/u-boot' bsp

这个是由于 rk3588s-rock-5a-u-boot.dtsi 加了太多节点导致的

nascs commented on arch/arm/dts/rockchip/rk3588s-radxa-cm5-io-u-boot.dtsi:
```diff
@@ -0,0 +1,280 @@
+ */
+
+#include "rk3588-u-boot.dtsi"
+
```
把 rk3588s-radxa-cm5-io-u-boot.dtsi 放到 arch/arm/dts/rockchip/ 目录下， rk3588s-radxa-cm5-io-u-boot.dtsi 就不会自动包含上一级的 rk3588s-radxa-cm5-io.dts 了， 要写 Makefile 之类的嘛？

nascs replied:
在 rk3588s-radxa-cm5-io-u-boot.dtsi 这里去 include "../rk3588s-radxa-cm5-io.dts" 也不行， 编译会报错，提示找不到一些节点

RadxaYuntian replied:
rk3588s-radxa-cm5-io-u-boot.dtsi肯定是跟 rk3588s-radxa-cm5-io.dts一样都放在arch/arm/dts，不然代码怎么找到设备树？但他们引用的内核dtsi就可以放到arch/arm/dts/rockchip/
然后你的那个给内核适配的rk3588-u-boot.dtsi也放到arch/arm/dts/rockchip/这里

nascs replied:
是 rk3588-linux.dtsi 吧， 内核没有 rk3588-u-boot.dtsi

RadxaYuntian replied:
你的rk3588s-radxa-cm5-io-u-boot.dtsi难道不会引用rk3588-u-boot.dtsi？

nascs replied:
会引用 rk3588-u-boot.dtsi， 但是这个文件原来就在 arch/arm64/dts/ 目录下， 并没有 "给内核适配的rk3588-u-boot.dtsi" 吧

RadxaYuntian replied:
https://github.com/radxa/u-boot/pull/109/files#diff-ff16864d36b672c6132a12a9d5a23f21b06531968825ab1e3e9351e4d823dc76
你这里改的东西其他的产品能用么？

nascs commented:
已经把原先 u-boot 没有的而 kernel 有的节点给放到 arch/ar/dts/rockchip/ 目录下了

nascs commented:
rk3588s-radxa-cm5-rpi-cm4-io 这次先不搞

RadxaYuntian commented on arch/arm/mach-rockchip/radxa_img.c:
```diff
@@ -1,55 +1,145 @@
-		free(buf);
-		return -ENOMEM;
-	}
+		if (strncasecmp(volume_label, "config", strlen("config")) == 0) {
```
你这个跟之前说的一样还是会匹配configA这种分区名啊

nascs replied:
> 你这个跟之前说的一样还是会匹配configA这种分区名啊

if (strncasecmp(volume_label, "config", strlen("config")) == 0 && strlen(volume_label) == strlen("config")) {
刚刚在本地写了个 demo, 验证这样不会匹配到 "configA" 这类字符

RadxaYuntian replied:
你这就等于没有n的strcasecmp

RadxaYuntian reviewed (changes requested):
另外你现在改了之后有测试现有的产品还能编译通过么？

nascs commented:
> 另外你现在改了之后有测试现有的产品还能编译通过么？

bsp 那边的两个 patch 要删掉， 部分产品会有冲突， https://github.com/radxa-repo/bsp/pull/123

RadxaYuntian commented:
首先那个PR已经说了你的U-Boot的profile是错的，那两个分支没有你这里的改动，需要保留那里的补丁。
还有就是这些[从内核来的dtsi要放到rockchip目录里面](https://github.com/radxa/u-boot/pull/109#pullrequestreview-2467149383)，这样我们现有的产品的设备树没有变化，如果rockchip之后要更新这里的内容也方便合并

nascs commented:
> 另外你现在改了之后有测试现有的产品还能编译通过么？

现在全部都能编译了

nascs commented:
头文件是不是也需要隔离？

nascs commented on include/dt-bindings/clock/rk3588-cru.h:
```diff
@@ -684,8 +684,6 @@
 #define TCLK_PMU1WDT			689
 #define CLK_CR_PARA			690
-#define CLK_GMAC0			691
-#define CLK_GMAC1			692
```
这里会有些冲突

RadxaYuntian commented on include/dt-bindings/thermal/thermal.h:
这个复制感觉没什么意义

RadxaYuntian commented on arch/arm/dts/rk3588s-u-boot.dtsi:
这个干脆也全部移到rockchip/rk3588s-u-boot.dtsi里面好了，外面只有我们板级的dts

RadxaYuntian reviewed (changes requested):
另外我们现在两个板级的dts和内核相比是不是只有include路径有一些变化？别的代码都没改吧？

nascs commented:
> 另外我们现在两个板级的dts和内核相比是不是只有include路径有一些变化？别的代码都没改吧？

是的

RadxaYuntian reviewed (changes requested):
提交整理一下吧

nascs committed:
include: port necessary header file from kernel

Sync from kernel tree, https://github.com/radxa/kernel/tree/linux-6.1-stan-rkr4.1
Commit: 3bf163742624 (drm/sched: Use drm sched lockdep map for submit_wq)

Signed-off-by: Nascs Fang <nascs@radxa.com>

nascs committed:
dts: rockchip: port rk3588s from kernel

Sync from kernel tree, https://github.com/radxa/kernel/tree/linux-6.1-stan-rkr4.1
Commit: 3bf163742624 (drm/sched: Use drm sched lockdep map for submit_wq)

Signed-off-by: Nascs Fang <nascs@radxa.com>

nascs committed:
dts: rockchip: rk3588: add rk3588 u-boot dtsi

Signed-off-by: Nascs Fang <nascs@radxa.com>

nascs committed:
dts: rockchip: rk3588s: add rk3588s u-boot dtsi

Signed-off-by: Nascs Fang <nascs@radxa.com>

nascs committed:
dts: radxa cm5 io: port device tree files from kernel

Sync from kernel tree, https://github.com/radxa/kernel/tree/linux-6.1-stan-rkr4.1
Commit: 3bf163742624 (drm/sched: Use drm sched lockdep map for submit_wq)

Signed-off-by: Nascs Fang <nascs@radxa.com>

nascs committed:
configs: radxa cm5 io: reorganise defconfig with make savedefconfig

Signed-off-by: Nascs Fang <nascs@radxa.com>

nascs committed:
dts: radxa cm5 io: add radxa cm5/cm5 io u-boot dtsi

Signed-off-by: Nascs Fang <nascs@radxa.com>

nascs committed:
arch: arm: mach-rockchip: load u-boot logo from partition with volume label "config"

Previously, the U-Boot logo (BMP file) was loaded from the first partition.
This commit changes the behavior to load the logo from a partition with the label "config" or "CONFIG".

Signed-off-by: Nascs Fang <nascs@radxa.com>

nascs commented:
rk2410, 全部编译通过

RadxaStephen commented:
验证可用。

再修改一下：把几个提交的 git message 的标题和正文，用空行隔开。

RadxaStephen commented:
Merged.
